torch/csrc/autograd/saved_variable.cpp

```diff
@@ -78,11 +78,15 @@ Variable SavedVariable::unpack(std::shared_ptr<Function> saved_for) const {
   // NB: saved views are unpacked as normal Variables (not views) even though
   // they still share the same storage. This works only because we never call
   // in-place functions on unpacked variables.
-  Variable var;
+  Variable var = as_variable_ref(data_).detach();
```

Comment:
TODO: It seems unclear whether we have invariant that the SavedVariable will only be unpacked once. (I tried removing `detach`/`variable_data` here and all tests pass.) If we do have that invariant, we should always use `detach`/`variable_data` here.

Reply:
how would you figure out if we have that invariant or not?

Reply:
In the following use case, a SavedVariable can be unpacked more than once:
```python
>>> a = torch.zeros(2, 2).fill_(3).requires_grad_()
>>> b = torch.zeros(2, 2).fill_(5).requires_grad_()
>>> c = a*b
# SavedVariable being saved:  3  3
#  3  3
# [ CPUFloatType{2,2} ]
# SavedVariable being saved:  5  5
#  5  5
# [ CPUFloatType{2,2} ]

>>> d = c.sum()
>>> d.backward(retain_graph=True)
# SavedVariable being unpacked:  3  3
#  3  3
# [ CPUFloatType{2,2} ]
# SavedVariable being unpacked:  5  5
#  5  5
# [ CPUFloatType{2,2} ]

>>> d.backward()
# SavedVariable being unpacked:  3  3
#  3  3
# [ CPUFloatType{2,2} ]
# SavedVariable being unpacked:  5  5
#  5  5
# [ CPUFloatType{2,2} ] 
```
In the above example, `a` and `b` are unpacked more than once, hence we should always use `detach`/`variable_data` to make a shallow-copy of the SavedVariable here, so that each unpacking does not affect the original SavedVariable.